src/main/java/com/multithreading/TextProcessor.java

```diff
@@ -1,38 +1,54 @@
 package com.multithreading;
 
+import java.util.concurrent.atomic.AtomicInteger;
+
 class Capitalize implements Runnable {
     private final char[] chars;
     private final int index;
+    private final AtomicInteger numTasksLeft;
 
-    public Capitalize(char[] chars, int index) {
+    public Capitalize(char[] chars, int index, AtomicInteger numTasksLeft) {
         this.chars = chars;
         this.index = index;
+        this.numTasksLeft = numTasksLeft;
     }
 
     @Override
     public void run() {
         synchronized (chars) {
             chars[index] = Character.toUpperCase(chars[index]);
         }
+        synchronized (numTasksLeft) {
+            numTasksLeft.decrementAndGet();
+            numTasksLeft.notifyAll();
+        }
     }
 }
 
 public class TextProcessor {
 
-    public String process(String input) throws InterruptedException {
-        Thread[] threads = new Thread[input.length()];
+    private final ThreadPool threadPool;
 
+    public TextProcessor(int numThreads) {
+        this.threadPool = new ThreadPool(numThreads);
+    }
+
+    public String process(String input) throws InterruptedException {
         char[] chars = input.toCharArray();
+        AtomicInteger numTasksLeft = new AtomicInteger(input.length());
 
         for (int i = 0; i < chars.length; i++) {
-            threads[i] = new Thread(new Capitalize(chars, i));
-            threads[i].start();
+            threadPool.submit(new Capitalize(chars, i, numTasksLeft));
         }
 
-        for (Thread thread : threads) {
-            thread.join();
+        synchronized (numTasksLeft) {
+            while (numTasksLeft.get() > 0) {
```

Comment:
my eyes... my eyes.... Keeping a while loop with wait() in a synchronized block is hurting!

Also the whole point of atomic integers is you don't have to synchronize them so much.
Since your basic approach had a similar problem, how about tracking the tasks as a set of tasks Ids, when a task is completed, it notifies its id is completed and you remove it from the set. when the set is empty you are done.
Some benefit (vs counter) are:
- unless a task can guess the id of another task they can't accidentally decrement more than once
- you can add additional properties for each task (using a map<taskId, props> instead of set) like how long its allowed to run for for example and when it started. so you can have a clean-up job/watchdog that instructs long tasks to die.

Reply:
sorry for hurting your eyes :") but i'm kinda stuck here: based on my understanding there are a few restrictions:
* runnable/thread cannot return anything
* each runnable/thread should not be aware of anything more than mutex, its own character and task id
* the main thread can only wait on 1 object at a time - so if each thread notifies its own task ID object, it can't be guaranteed to be collected
i might have some wrong understandings here, but I'm kinda out of clues on how to let the main thread be aware of which task ID is notified while sticking with the above restrictions

Reply:
You can definitely pass a callable which is one of the approaches i've suggested this in the basic PR.

Main thread doesn't need to wait on one thread. It should wait for tasks to be completed then feed another task to the same thread when it's free. If a task cannot run for too long and is interruptible (based on tasks properties from the map i suggested) you could interrupt the thread and discard it, start a new thread for your pool.